src/language_concepts/generics.md

```diff
@@ -0,0 +1,78 @@
+# Generics
+
+Generics allow you to use the same functions with multiple different concrete data types. You can read more about the concept of generics in the Rust documentation [here](https://doc.rust-lang.org/book/ch10-01-syntax.html).
+
+Here is a trivial example showing the identity function that supports any type. In Rust, it is common to refer to the most general type as `T`. We follow the same convention in Noir.
+
+```rust noplaypen
+fn id<T>(x: T) -> T  {
+    x
+}
+```
+
+## Arbitrary length arrays
+
+This concept is more useful for more specific types, like number generics. These can be used to define a function that iterates over an array of arbitrary length.
+
+```rust noplaypen
+struct RepeatedValue<T> {
+    value: T,
+    count: comptime Field,
+}
+
+impl<T> RepeatedValue<T> {
+    fn new(value: T) -> Self {
+        Self { value, count: 1 }
+    }
+
+    fn increment(mut repeated: Self) -> Self {
+        repeated.count += 1;
+        repeated
+    }
+
+    fn print(self) {
+        for _i in 0 .. self.count {
+            dep::std::println(self.value);
+        }
+    }
+}
+
+fn main() {
+    let mut repeated = RepeatedValue::new("Hello!");
+    repeated = repeated.increment();
+    repeated.print();
+}
+```
+
+The `print` function will print `Hello!` an arbitrary number of times, twice in this case.
+
+## Function Generics
+
+It can also be useful to create generic functions that will work over all types, `T`.
+
+For example, how can we write a generic equality function for arrays of any type, `T`?
```

Comment:
The example below is meant to show how we can write a function that is generic over all T for which some function (in this case elem_eq) is available. So the description of working over all types T isn't accurate. We could say:

```
## Calling functions on generic parameters

Unlike Rust, Noir does not have traits, so how can one translate the equivalent 
of a trait bound in rust into Noir? That is, how can we write a function that is 
generic over some type `T`, while also requiring there is a function like `
eq: fn(T, T) -> bool` that works on the type? The answer is that we can
 translate this by passing in the function manually. Here's an example of 
implementing array equality in Noir:
```